Make regex field optional

psarka opened the pull request:
This PR makes the regex field optional.

The issues of potentially [illegal path](https://github.com/ContinuumIO/intake/issues/265) or file name collisions are not affected by this PR.

Alleviates https://github.com/ContinuumIO/intake/issues/254. It does not fully resolve it, as if user does provide a regex which does not match, files are still dropped in the cwd instead of `cache_dir`.

martindurant commented:
As it happens, this seems to make an illegal path on windows

psarka commented:
oh dear :man_facepalming:

martindurant commented on docs/source/catalog.rst:
```diff
@@ -287,6 +285,11 @@ The special value ``"catdir"`` implies that cached files will appear in the same
 
+Optionally, the cache section can have a ``regex`` attribute, that modifies the path of the cache on
+the disk. By default, the cache path is made by concatenating ``cache_dir``, dataset name, hash of
+the url, and the url itself (without the protocol). ``regex`` attribute allows to remove part of the
```
Is there a better description of regex than this? It feels a little hard to understand - but indeed, the usage *is* hard to uderstand.

psarka replied:
I was not able to come up with something clear and succinct, so decided to explicitly spell out what it does. 

What could be a good idea, is to warn user about cache files ending up in cwd if regex is not well specified. Where well specified regex to my understanding is one that matches part of the url that includes the beginning, but excludes the protocol. Should I add this?

martindurant replied:
> warn user about cache files ending up in cwd

Sorry, I thought that would no longer be the case, that they would instead end up in the cache directory, under the hash, but with a deeper path

martindurant replied:
> Should I add this?

but yes, I think exactly that wording would be good

martindurant commented on intake/source/cache.py:
```diff
@@ -91,11 +91,15 @@ def _ensure_cache_dir(self):
+        if 'regex' in self._spec:
+            regex = r'%s' % sanitize_path(self._spec['regex'])
+        else:
+            regex = r'^'  # if no regex provided, use full path
```
Interesting minimum-code solution!

psarka replied:
:)

martindurant commented:
Fixes #254

psarka commented:
Instead of documenting accidental writing to cwd I realized I can fix it easily.

Note that to be sure that file always ends up in the cache dir I removed the `return url_path if cache_path == path logic`. I have to admit I don't understand why would `url_path` be useful as `cache_path`, and the only failing unit test (which I removed too), did not reveal the usefulness.

martindurant commented:
Fair enough. Could you please a test showing that files do not end up in the cwd for the failed regex case, but they do end up in the catalog directory path if so specified?

psarka commented:
I think this should test that. Can you please take a look?

martindurant commented:
Thanks for all your work and patience, @psarka
